diff --git a/src/js/emscripten-settings.ts b/src/js/emscripten-settings.ts
index b27fe5e7..dd74b7c0 100644
--- a/src/js/emscripten-settings.ts
+++ b/src/js/emscripten-settings.ts
@@ -69,8 +69,11 @@ export function createSettings(
     // library but it failed to find it in the file system. But for us that
     // means dependency resolution has already failed and we want to throw an
     // error anyways.
-    locateFile: (path: string) => config.indexURL + path,
-    instantiateWasm: getInstantiateWasmFunc(config.indexURL),
+    locateFile: (path: string) =>
+      path === "pyodide.asm.wasm" && config.wasmURL
+        ? config.wasmURL
+        : config.indexURL + path,
+    instantiateWasm: getInstantiateWasmFunc(config),
   };
   return settings;
 }
@@ -192,7 +195,7 @@ function getFileSystemInitializationFuncs(
 }
 
 function getInstantiateWasmFunc(
-  indexURL: string,
+  config: PyodideConfigWithDefaults,
 ): EmscriptenSettings["instantiateWasm"] {
   // @ts-ignore
   if (SOURCEMAP || typeof WasmOffsetConverter !== "undefined") {
@@ -205,7 +208,10 @@ function getInstantiateWasmFunc(
     // typeof WasmOffsetConverter !== "undefined" checks for asan.
     return;
   }
-  const { binary, response } = getBinaryResponse(indexURL + "pyodide.asm.wasm");
+  const wasmURL = !!config.wasmURL
+    ? config.wasmURL
+    : config.indexURL + "pyodide.asm.wasm";
+  const { binary, response } = getBinaryResponse(wasmURL);
   const jsvErrorImportPromise = getJsvErrorImport();
   return function (
     imports: { [key: string]: { [key: string]: any } },
diff --git a/src/js/pyodide.ts b/src/js/pyodide.ts
index 297380c5..580e4d98 100644
--- a/src/js/pyodide.ts
+++ b/src/js/pyodide.ts
@@ -109,6 +109,16 @@ export interface PyodideConfig {
    * Default: ```${indexURL}/python_stdlib.zip```
    */
   stdLibURL?: string;
+  /**
+   * The URL from which to load the ``pyodide.asm.mjs`` glue script.
+   * Default: ```${indexURL}/pyodide.asm.mjs```
+   */
+  asmURL?: string;
+  /**
+   * The URL from which to load the ``pyodide.asm.wasm`` binary.
+   * Default: ```${indexURL}/pyodide.asm.wasm```
+   */
+  wasmURL?: string;
   /**
    * Override the standard input callback. Should ask the user for one line of
    * input. The :js:func:`pyodide.setStdin` function is more flexible and
@@ -339,7 +349,7 @@ async function loadWasmScript(
     return config.createPyodideModule;
   }
 
-  const scriptSrc = `${config.indexURL}pyodide.asm.mjs`;
+  const scriptSrc = config.asmURL || `${config.indexURL}pyodide.asm.mjs`;
   return (await loadScript(scriptSrc)).default;
 }
 
